synapse/event_auth.py

```diff
@@ -729,16 +730,36 @@ def _check_power_levels(
 ) -> None:
     user_list = event.content.get("users", {})
     # Validate users
-    for k, v in user_list.items():
+    for k in user_list.keys():
         try:
             UserID.from_string(k)
         except Exception:
             raise SynapseError(400, "Not a valid user_id: %s" % (k,))
 
-        try:
-            int(v)
-        except Exception:
-            raise SynapseError(400, "Not a valid power level: %s" % (v,))
+    # Reject events with stringy power levels if required by room version
+    if (
+        event.type == EventTypes.PowerLevels
+        and room_version_obj.msc3667_int_only_power_levels
+    ):
+        for k, v in event.content.items():
+            if k in {
+                "users_default",
+                "events_default",
+                "state_default",
+                "ban",
+                "redact",
+                "kick",
+                "invite",
+            }:
+                if not isinstance(v, int):
+                    raise AuthError(403, f"{v} must be an integer.")
```

Comment:
The existing validation in `_check_power_levels` raises `SynapseError(400)`s. I think we want to do that too for the new validation?

Reply:
Good call, I have fixed.